iptables/table_test.go

```diff
@@ -535,6 +545,160 @@ func describeEmptyDataplaneTests(dataplaneMode string) {
+				{Name: "non-cali-chain", Rules: []Rule{
+					{Action: AcceptAction{}, Comment: "non-cali 1"},
+					{Action: DropAction{}, Comment: "non-cali 2"},
+				}},
+				{Name: "cali-foobar", Rules: []Rule{
+					{Action: AcceptAction{}, Comment: "cali 1"},
+					{Action: DropAction{}, Comment: "cali 2"},
+				}},
+			})
+			table.Apply()
+		})
+		It("should be in the dataplane", func() {
+			Expect(dataplane.Chains).To(Equal(map[string][]string{
+				"FORWARD": {},
+				"INPUT":   {},
+				"OUTPUT":  {},
+				"non-cali-chain": {
+					"-m comment --comment \"cali:Z-OWODLe_LbHxmqg\" -m comment --comment \"non-cali 1\" --jump ACCEPT",
+					"-m comment --comment \"cali:tq-yEo1_1XQHZnMs\" -m comment --comment \"non-cali 2\" --jump DROP",
+				},
+				"cali-foobar": {
+					"-m comment --comment \"cali:cxE-1zsuD12R9YEG\" -m comment --comment \"cali 1\" --jump ACCEPT",
+					"-m comment --comment \"cali:1cpbPOGLTROlH4Sj\" -m comment --comment \"cali 2\" --jump DROP",
+				},
+			}))
+		})
+		Describe("then truncating the chain, with the iptables changed before iptables-restore", func() {
+			BeforeEach(func() {
+				dataplane.OnPreRestore = func() {
+					log.Warn("Simulating an insert in non-cali-chain before iptables-restore happens")
+					if chain, found := dataplane.Chains["non-cali-chain"]; found {
+						log.Warn("non-cali-chain exists; inserting random rule in non-cali-chain")
+						lines := prependLine(chain, "-j randomly-inserted-rule")
+						dataplane.Chains["non-cali-chain"] = lines
+					}
+				}
+
+				table.SetRuleInsertions("non-cali-chain", []Rule{
+					{Action: DropAction{}, Comment: "new drop rule"},
+				})
+				table.Apply()
+			})
+			It("should be updated", func() {
+				Expect(dataplane.Chains).To(Equal(map[string][]string{
+					"FORWARD": {},
+					"INPUT":   {},
+					"OUTPUT":  {},
+					"non-cali-chain": {
+						"-m comment --comment \"cali:O9yEP97Dd2y-EskM\" -m comment --comment \"new drop rule\" --jump DROP",
+						"-j randomly-inserted-rule"},
+					"cali-foobar": {
+						"-m comment --comment \"cali:cxE-1zsuD12R9YEG\" -m comment --comment \"cali 1\" --jump ACCEPT",
+						"-m comment --comment \"cali:1cpbPOGLTROlH4Sj\" -m comment --comment \"cali 2\" --jump DROP",
+					},
+				}))
+			})
+		})
+	})
+
+	Describe("inserting into a non-Calico chain results in the expected writes", func() {
+		BeforeEach(func() {
+			table.SetRuleInsertions("FORWARD", []Rule{
+				{Action: DropAction{}, Comment: "a drop rule"},
+				{Action: AcceptAction{}, Comment: "an accept rule"},
+			})
+			table.Apply()
+		})
+		It("should update the dataplane", func() {
+			Expect(dataplane.Chains).To(Equal(map[string][]string{
+				"FORWARD": {
+					"-m comment --comment \"cali:upaItQyFMdN7MTTl\" -m comment --comment \"a drop rule\" --jump DROP",
+					"-m comment --comment \"cali:EpCg3AYNp_DftVFS\" -m comment --comment \"an accept rule\" --jump ACCEPT",
+				},
+				"INPUT":  {},
+				"OUTPUT": {},
+			}))
+
+			expectedWrites := set.New()
```

Comment:
I think, to make the chainMods assertion useful, it needs to make sure that there aren't any unexpected writes.  (Right now, you're only verifying tha thte expected writes _were_ done, not that there weren't any others.)

However, you've already covered that base by doing this check: 

```Expect(dataplane.CmdNames).To(ConsistOf("iptables", "iptables-nft-save"))```

There can't be an unexpected write if the code didn't call `iptables-restore` at all!

Since chainMods are a bit hard to pin down for inserts, I say drop the chainMod assertions and rely on the check that iptables-restore didn't get called.